### PR TITLE
Add Initial Documentation, Tests, and Examples

### LINKED ISSUE #2: Reorganize modules
Data.Validation should be split up.
I think 2 modules are needed, one for the main stuff, and another for what you have labeled as Generic validations and Validation helpers in the export list.

### LINKED ISSUE #4: Call For General Validators
What general validators should we add to the lib? Here's my list so far:

 - IsEqual - using the `Eq` type class.
 - IsNotEqual - using the `Eq` type class.
 - isLessThan - using the `Ord` type class.
 - isGreaterThan - using the `Ord` type class.
 - matchesRegex - regular expression checking.
 - hasElem - based on `elem`.
 - doesNotHaveElem - based on `elem`.
 - hasAll - checks every element in a foldable.
 - hasAny - check every element in a foldable.

### PR DESCRIPTION
**Pull request by devinlyons**
Closes #2, closes #4.

### CODE REVIEW

File: src/Data/Validation.hs [hunk 1]
```diff
@@ -47,184 +23,274 @@ import Data.Map hiding (null)
...
+  data MyFailures = EmptyEmailAddress | MalformedEmailAddress 
+  validateEmailAddress :: String -> Either MyFailures EmailAddress
+  @
+
+  In this case:
+  * `a` ~ String
+  * `b` ~ EmailAddress
+  * `c` ~ MyFailures
+
+  The transformation from `a` to `b` is imporant and provides a type safe way to proove that validation was successful.
```
Comment (sambolias) [inline src/Data/Validation.hs]: *prove

File: src/Data/Validation.hs [hunk 2]
```diff
@@ -47,184 +23,274 @@ import Data.Map hiding (null)
...
+  However, rather than using the `Either` type, this library uses the 'Proof' type.
+  A 'Proof' represents either a validated type or a collection of failures.
+  Notice, we use the term validation _failures_ instead of _errors_ 
+  to deferentiate between validation and error handling.
```
Comment (sambolias) [inline src/Data/Validation.hs]: *differentiate

File: src/Data/Validation.hs (hunk 1 above)
Comment (sambolias) [reply src/Data/Validation.hs]: *important

File: src/Data/Validation.hs [hunk 3]
```diff
@@ -47,184 +23,274 @@ import Data.Map hiding (null)
...
+-- | Adds a validation failure to the result and continues validation.
+dispute :: ValueCtx a -> f -> VCtx f (ValueCtx a)
+dispute v@(Field n _) f = DisputedCtx [] (singleton [n] [f]) v
+dispute v@(Global _) f  = DisputedCtx [f] empty v
 
--- Standard validation functions can be created to make it easier to perform common validations.
-disputeWith :: (a -> Maybe e) -> ValueCtx a -> VCtx e (ValueCtx a)
-disputeWith f v = case f (getValue v) of
-  Just e  -> dispute v e
+-- | Performs a validation using a given function and handls the result.
```
Comment (sambolias) [inline src/Data/Validation.hs]: to handle?

File: src/Data/Validation.hs [hunk 4]
```diff
@@ -47,184 +23,274 @@ import Data.Map hiding (null)
...
 
-refuteWithFact :: e -> (a -> Bool) -> ValueCtx a -> VCtx e (ValueCtx a)
-refuteWithFact e f = refuteWith (\a -> bool (Left e) (Right a) $ f a)
+-- | Similar to 'refuteWith' except that the given failure is added if the given function returns False.
+-- This function does not transform the value like 'refuteWith'.
+refuteWithFact :: f -> (a -> Bool) -> ValueCtx a -> VCtx f (ValueCtx a)
+refuteWithFact f fn = refuteWith (\a -> bool (Left f) (Right a) $ fn a)
 
-refuteWithProof :: (a -> Proof e b) -> ValueCtx a -> VCtx e (ValueCtx b)
+-- | Performs a validation using a given function and handls the result.
```
Comment (sambolias) [inline src/Data/Validation.hs]: same

File: src/Data/Validation.hs [hunk 5]
```diff
@@ -47,184 +23,274 @@ import Data.Map hiding (null)
...
+maxValue l f = disputeWith $ bool (Just f) Nothing . (>=) l
+
+-- | Checks that two fields are equal.
+-- If not, it adds the given failure to the result and validation continues.
+isMatch :: Eq a => f -> VCtx f a -> ValueCtx a -> VCtx f (ValueCtx a)
+isMatch f (ValidCtx a)        = disputeWith (testMatch f a)
+isMatch f (DisputedCtx _ _ a) = disputeWith (testMatch f a)
+isMatch f (RefutedCtx _ _)    = pure
+
+-- | Vaidates a value that implements 'Validatable' and includes any failures under the parent field.
```
Comment (sambolias) [inline src/Data/Validation.hs]: *Validates
and should this mention something about how it determines the field?

**Review by sambolias — CHANGES_REQUESTED** [review]
I like this organization, this lib is small enough to not need the extra level of separating generic validators from main validation module. And the docs look descriptive enough, nice compliment to RFC docs. My comments are mostly for small typos

**Comment by devinlyons** [comment]
@sambolias, let's proof read later. There's a lot of stuff that I am going to be pulling from the RFC to populate the docs. There are likely to be some major edits in the near future.

**Comment by sambolias** [comment]
@devinlyons should I approve based solely on the reorganization then? Just want to make sure I reviewed what you wanted.

**Comment by devinlyons** [comment]
Don't approve anything. This is more of a discussion for now. I should probably have opened this as a draft PR.

File: examples/Examples/Data/Primitives.hs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  show EmailError = "bad email"
+  show RequiredError = "required field missing value"
+  show OtherError = "something else wrong"
+
+-- app specific types
+data UsernameTag
+type Username = Tagged UsernameTag String
+mkUsername :: String -> Proof MyError Username
+mkUsername s = fromVCtx $ withValue s (isNotNull EmailError)
+  >>= pure . Tagged
```
Comment (boj) [inline examples/Examples/Data/Primitives.hs]: Please prefer do-notation where possible.

File: src/Data/Validation.hs (hunk 1 above)
Comment (devinlyons) [reply src/Data/Validation.hs]: Fixed.

File: src/Data/Validation.hs (hunk 2 above)
Comment (devinlyons) [reply src/Data/Validation.hs]: Fixed.

File: src/Data/Validation.hs (hunk 4 above)
Comment (devinlyons) [reply src/Data/Validation.hs]: This seems right to me.

File: src/Data/Validation.hs (hunk 5 above)
Comment (devinlyons) [reply src/Data/Validation.hs]: Fixed.

File: examples/Examples/Data/Primitives.hs (hunk 6 above)
Comment (devinlyons) [reply examples/Examples/Data/Primitives.hs]: Done.

**Review by boj — APPROVED** [review]
Documentation looks good.

File: data-validation.cabal [hunk 7]
```diff
@@ -16,13 +16,38 @@ maintainer:
 category:            Data
 extra-source-files:  CHANGELOG.md, README.md
 
+library examples-internal
+  exposed-modules:    Examples.Data.ComplexTypes
+                    , Examples.Data.Primitives
+  -- other-modules:      
+  -- other-extensions:
```
Comment (sambolias) [inline data-validation.cabal]: probably want to remove comments

File: data-validation.cabal [hunk 8]
```diff
@@ -16,13 +16,38 @@ maintainer:
...
+  hs-source-dirs:     examples
+  default-language:   Haskell2010
+  ghc-options:       -Wall -j -O0 -v0
+
 library
   exposed-modules:    Data.Validation
                     , Data.Validation.Transforms
-  -- other-modules:
+  other-modules:      Data.Validation.Internal
   -- other-extensions:
```
Comment (sambolias) [inline data-validation.cabal]: remove comment

File: examples/Examples/Data/Primitives.hs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  show EmailError = "bad email"
+  show RequiredError = "required field missing value"
+  show OtherError = "something else wrong"
+
+-- app specific types
+newtype Username = Username { unUsername :: String }
+  deriving Show
+mkUsername :: String -> Proof MyError Username
+mkUsername s = fromVCtx $ do 
+  v <- withValue s (isNotNull EmailError)
```
Comment (sambolias) [inline examples/Examples/Data/Primitives.hs]: This should be `NameError`

File: src/Data/Validation/Internal.hs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+import Language.Haskell.TH (Name)
+------------------------------------------------------------------------------------------------------------------------
+
+-- | A type that holds aggregated validation failures.
+data VCtx f a 
+  -- | A value that is assumed to be valid.
+  = ValidCtx a 
+  -- | A value that has failures but can continue to be validated.
+  | DisputedCtx [f] (Map [Name] [f]) a 
+  -- | A that has failures and cannot be validated further.
```
Comment (sambolias) [inline src/Data/Validation/Internal.hs]: *A value that*

File: examples/Examples/Data/Primitives.hs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+, PhoneNumber
+, mkPhoneNumber
+, unPhoneNumber
+) where
+
+------------------------------------------------------------------------------------------------------------------------
+import Data.Validation
+------------------------------------------------------------------------------------------------------------------------
+
+data MyError
```
Comment (sambolias) [inline examples/Examples/Data/Primitives.hs]: Too keep with vocabulary in RFC this should probably be `ValidationFailures` or similar

**Review by sambolias — CHANGES_REQUESTED** [review]
Added some minor change comments, but overall looks great. Testing seems thorough for helpers, but should you also test some other aspects, like combining `VCtx`s with any of the instances defined in Internal.hs?

File: src/Data/Validation/Internal.hs (hunk 10 above)
Comment (devinlyons) [reply src/Data/Validation/Internal.hs]: I'm not sure what you are trying to point out here.

Comment (boj) [reply src/Data/Validation/Internal.hs]: "A that has" -> "A value that has"

Comment (devinlyons) [reply src/Data/Validation/Internal.hs]: Oh, wow. I see it now...

File: src/Data/Validation.hs [hunk 12]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+  * @a@ ~ 'String'
+  * @b@ ~ EmailAddress
+  * @c@ ~ MyFailures
+
+-}
+
+{- $proof
+
+  The transformation from `a` to `b` is important and provides a type safe way to prove that validation was successful.
+  However, rather than using the 'Either' type, this library uses the 'Proof' type.
```
Comment (sambolias) [inline src/Data/Validation.hs:115]: "However" seems out of place here

File: src/Data/Validation.hs [hunk 13]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+  Field failures are useful for identifying a specific field in a record that is invalid.
+  Fields are identified using a list of 'Language.Haskell.TH.Name' types.
+  There are two ways to create this type: the @TemplateHaskellQuotes@ extension and the 'Language.Haskell.TH.mkName' function.
+
+  Using the @TemplateHaskellQuotes@ language extension, you can easily create 'Language.Haskell.TH.Name's using a special syntax.
+  For a records like:
+
+  > data User = User { emailAddress :: String }
+
+  The name can be retrived by referecing the name with a single quote in front:
```
Comment (sambolias) [inline src/Data/Validation.hs]: *retrieved by *referencing

File: src/Data/Validation.hs [hunk 14]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+  This method does generate a fully qualified name that includes the module name.
+  The base name can be accessed using the 'Language.Haskell.TH.nameBase' function.
+  This extension is considered safe Haskell while the @TemplateHaskell@ extension is not.
+  See <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/template_haskell.html#syntax> for details.
+
+  Another approach is to use the 'Language.Haskell.TH.mkName' function.
+  This allows field names to be generated for non-record types.
+  However, it does require managing magic strings.
+
+  To finish our discusion on the 'Proof' type, the reason the key is a list of names is because of subfields.
```
Comment (sambolias) [inline src/Data/Validation.hs]: *discussion

File: src/Data/Validation.hs [hunk 15]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
-fromVCtx (DisputedCtx ges les _) = Invalid ges les
-fromVCtx (RefutedCtx ges les)    = Invalid ges les
+fromVCtx (DisputedCtx gfs lfs _) = Invalid gfs lfs
+fromVCtx (RefutedCtx gfs lfs)    = Invalid gfs lfs
 
-class Validatable e a b | a -> e b where
-  validation :: a -> VCtx e b
+{- $valuectx
+  This library uses composable validations.
+  As such, we need a type that carries information about the thing being validation.
```
Comment (sambolias) [inline src/Data/Validation.hs]: * thing being validated

File: src/Data/Validation.hs [hunk 16]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+  Specifically, we need a type to carry its value and, optionaly, its name.
```
Comment (sambolias) [inline src/Data/Validation.hs]: * optionally

File: src/Data/Validation.hs [hunk 17]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+withField n a fn = fn (Field n a) >>= pure . getValue
+
+-- | Performs some given validation using a 'Global' with a given value.
+withValue :: a -> (ValueCtx a -> VCtx f (ValueCtx b)) -> VCtx f b
+withValue a fn = fn (Global a) >>= pure . getValue
+
+{- $primitives
+  Validating primitives is a special case because the validated type and unvalidated type are often the same.
+  An email address is just a specially formatted string, so both the unvalidated email address and the validated one have the same types.
+  To get around this, you can wrap a primitaive in a `newtype`, hide the constructor, and create a smart constructor that validates the input.
```
Comment (sambolias) [inline src/Data/Validation.hs]: *primitive

File: src/Data/Validation.hs [hunk 18]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+  This allows for a kind of type safty known as `correct by construction`.
```
Comment (sambolias) [inline src/Data/Validation.hs]: *safety

File: src/Data/Validation.hs [hunk 19]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+  
+  mkEmailAddress :: String -> Proof MyFailures EmailAddress
+  mkEmailAddress s = fromVCtx $ do              -- (1)
+    v \<- withValue s $ \\v -> do                 -- (2)
+      isNotNull EmptyEmail v                    -- (3)
+      disputeWithFact InvalidEmail (elem '@') v -- (4)
+    return $ EmailAddress v                     -- (5)
+  @
+
+  Starting with line (2), the 'withValue' function is used to create a 'ValueCtx' using string passed into the funciton.
```
Comment (sambolias) [inline src/Data/Validation.hs]: * function

File: src/Data/Validation.hs [hunk 20]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+  The lambda function that follows takes the 'ValueCtx' and runs it through several validators using `do` syntax.
+  This is possible because 'VCtx' has a 'Monad' instance.
+  
+  Line (1) trasforms the result from a 'VCtx' to a 'Proof' as discussed above.
```
Comment (sambolias) [inline src/Data/Validation.hs]: * transforms

File: src/Data/Validation.hs [hunk 21]
```diff
@@ -41,190 +76,581 @@ module Data.Validation
...
+    then disputeMany v fs
+    else pure v
+
+-- | Every element must be valid.
+ifAll :: (a -> Maybe f) -> ValueCtx [a] -> VCtx f (ValueCtx [a])
+ifAll fn v = case catMaybes . fmap fn $ getValue v of 
+  [] -> pure v
+  fs -> disputeMany v fs
+
+-- | Validate each element with a given funciton.
```
Comment (sambolias) [inline src/Data/Validation.hs]: * function

**Review by sambolias — CHANGES_REQUESTED** [review]
I missed a file on first pass, only found small typos

File: src/Data/Validation.hs (hunk 12 above)
Comment (devinlyons) [reply src/Data/Validation.hs:115]: This is right. I spent the section above extolling the virtues of the `Either` type. Then, this sentence says "yeah, we're not actually going to use that either type..."

Comment (sambolias) [reply src/Data/Validation.hs:115]: I see, it is an alternative to the example. I think the rather alone would suffice, but this seems OK, if you like it better that's good enough for me.

**Review by sambolias — APPROVED** [review]

**Pull request merged**